design/ceph/osd-migration.md

```diff
@@ -0,0 +1,188 @@
+
+**NOTE**: Once the OSDs are migrated to a new backend store, say, `bluestore-rdr`, they won't be allowed to be migrated back to the legacy store (bluestore).
+
+- Add a new label `osd-store:<osd store type>` to all the OSD pods.
+- This label will help to identify the current backend store being used for the OSD and enable the operator to decide if the OSD should be migrated in case the user changes the backend store type in the spec (`spec.Storage.store.type`).
+
+### New OSDs
+
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the OSDs.
+
+#### OSD Prepare
+
+- The `ceph-volume prepare` command is updated to use backend store provided in the `spec.storage.store`. If no value is provided for `spec.storage.store` then default to `bluestore` backend.
+
+    - RAW MODE
+        ```
+        ceph-volume raw prepare <OSD_STORE_FLAG> --data /dev/vda
+        ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm prepare <OSD_STORE_FLAG> --data /dev/vda
+      ```
+
+#### OSD Create
+
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.store`. If no values are provided by the user, then `OSD_STORE_FLAG` defaults to `bluestore`
+- `ceph-volume` commands use the `OSD_STORE_FLAG` while activating the OSDs.
+
+    - RAW MODE
+      ```
+      ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs
+      ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"
+      ```
+
+**NOTE**: ceph-volume `activate` command doesn't require `OSD_STORE_FLAG` to be passed as an argument. It auto detects the backend store that was used during the OSD `prepare`
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a Ceph version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  store:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### Replace OSD
+- Operator reconciler will replace one OSD at a time.
+- A config map will be used to store the list of OSDs to be replaced and the OSD being replaced currently.
+- OSD replacement steps:
+  1. List out all OSDs where `osd-store:<osd store type>` does not match with `spec.storage.store.type`.
```

Comment:
The OSDs from step 1 will only be considered for this transition if the transition is supported, right? For example, if all OSDs are converted to `bluestore_rdr,` then the cluster CR is changed back to `bluestore`, we will just not consider them for upgrade and perhaps log a warning, right?

Reply:
Yes. Rook won't support converting OSDs from `bluestore-rdr` to `bluestore`.  There is a not above (`Once the OSDs are migrated to a new backend store,...` ) with this mention.